numerics/concepts.hpp

```diff
@@ -0,0 +1,37 @@
+#pragma once
+
+#include <concepts>
+#include <type_traits>
+
+namespace principia {
+namespace numerics {
+namespace _concepts {
+namespace internal {
+
+template<typename T>
+concept one_dimensional = requires(T& t, int index) {
+  { t[index] } -> std::same_as<typename T::Scalar&>;
+};
+
+template<typename T>
+concept two_dimensional = requires(T& t, int row, int column) {
+  { t(row, column) } -> std::same_as<typename T::Scalar&>;
+};
+
+template<typename T>
+concept unbounded = requires(T& t) {
+  { t.size() } -> std::same_as<int>;
```

Comment:
I could imagine adding `size` to the fixed types, just like `std::array` has a size.

Reply:
Hmm, actually they `size`, `rows` and `columns`.  So I don't even need the `unbounded` concept, which is good because we want to handle fixed and unbounded similarly if possible.

Note that `size` probably doesn't make much sense for matrices, so I'll remove it ultimately.